### PR TITLE
[stm] fix regression introduced in de2d782

### PR DESCRIPTION
**Pull request by gares**
Alternative to #17742

### CODE REVIEW

**Comment by SkySkimmer** [comment]
@coqbot bench

**Comment by proux01** [comment]
Successfully tested on order.v in MathComp.

**Comment by Zimmi48** [comment]
Sure, I was planning an 8.17.1 release, but this fix is even a good justification in itself to do it.

**Comment by SkySkimmer** [comment]
Do we know why mathcomp is particularly impacted? Are there that much more notations than elsewhere?

**Comment by proux01** [comment]
> Do we know why mathcomp is particularly impacted? Are there that much more notations than elsewhere?

It's obviously linked to hierarchy-builder since the affected files are the heavy users of HB but I don't have a better understanding (not that even with the patch, those files remain memory heavy).

**Comment by proux01** [comment]
@coqbot run full ci

**Comment by gares** [comment]
> Do we know why mathcomp is particularly impacted? Are there that much more notations than elsewhere?

Elpi commands can import a module, so for now they are declared as `VtNow`.

**Comment by proux01** [comment]
Bench is good, CI failures are unrelated, let's merge.

@coqbot merge now

**Review by proux01 — APPROVED** [review]

**Comment by proux01** [comment]
@coqbot merge now

**Comment by Zimmi48** [comment]
Unfortunately, a naive backport of this PR doesn't produce code that builds, probably due to #17059, which also creates difficulties for backporting #17620. This creates the question whether backporting #17059 too would be worthwhile. Unfortunately, it changes quite a few signatures in the API of the STM. I don't know if these functions should be considered internal.

File: stm/stm.ml [hunk 1]
```diff
@@ -2695,7 +2695,7 @@ let process_transaction ~doc ?(newtip=Stateid.fresh ()) x c =
             begin match w with
               | VtNow ->
                 (* We need to execute to get the new parsing state *)
-                ignore(finish ~doc:dummy_doc);
+                let () = observe ~doc:dummy_doc (VCS.get_branch_pos (VCS.current_branch ())) in
```
Comment (SkySkimmer) [inline stm/stm.ml:2698]: `ignore (observe ...)` should work for the backport

**Pull request merged**